File: src/simulation_quantification.m
%% ========================================================================
%  SIMULATION_QUANTIFICATION.m - Simulation de la quantification
%  APP Signal - ISEP 2025-2026 - Problème III
%  ========================================================================
clear; close all; clc;

fprintf('=== Simulation de la quantification ===\n\n');

%% Paramètres
Fe = 44100;         % Hz
f0 = 1000;          % Fréquence du signal test (Hz)
V_max = 0.5;        % Amplitude crête (V)
A = 2 * V_max;      % Dynamique totale
duree = 0.1;        % Durée du signal (s)
bits_range = 1:20;  % Nombre de bits à tester

%% Génération du signal test
t = 0:1/Fe:(duree - 1/Fe);
N = length(t);

% Signal sinusoïdal
x = V_max * sin(2*pi*f0*t);

fprintf('Signal test : sinusoïde %.0f Hz, amplitude %.1f V, Fe = %d Hz\n', f0, V_max, Fe);
fprintf('Durée = %.3f s, N = %d échantillons\n\n', duree, N);

%% Quantification pour différents nombres de bits
SNR_mesure = zeros(1, length(bits_range));
SNR_theorie = zeros(1, length(bits_range));
debit = zeros(1, length(bits_range));

for idx = 1:length(bits_range)
    b = bits_range(idx);
    
    % Quantification
    x_q = round(x * 2^(b-1)) / 2^(b-1);
    
    % Erreur de quantification
    e = x - x_q;
    
    % Puissances
    P_signal = mean(x.^2);
    P_erreur = mean(e.^2);
    
    % SNR mesuré
    if P_erreur > 0
        SNR_mesure(idx) = 10*log10(P_signal / P_erreur);
    else
        SNR_mesure(idx) = Inf;
    end
    
    % SNR théorique
    q = A / (2^b);
    Pe_theorie = q^2 / 12;
    SNR_theorie(idx) = 10*log10(P_signal / Pe_theorie);
    
    % Débit
    debit(idx) = Fe * b * 2;  % stéréo
end

%% Affichage
figure('Name', 'Simulation de la quantification', 'Position', [100, 100, 1400, 900], 'Color', 'w');

% 1. SNR en fonction du nombre de bits
subplot(2,3,1);
plot(bits_range, SNR_mesure, 'bo-', 'LineWidth', 1.5, 'MarkerFaceColor', 'b');
hold on;
plot(bits_range, SNR_theorie, 'r--', 'LineWidth', 1.5);
yline(90, 'g--', 'Cible 90 dB', 'LineWidth', 1.5);
xlabel('Nombre de bits (b)'); ylabel('SNR (dB)');
title('SNR vs nombre de bits');
legend('Mesuré', 'Théorique', 'Location', 'southeast');
grid on;

% 2. SNR en fonction du débit
subplot(2,3,2);
plot(debit/1000, SNR_mesure, 'bo-', 'LineWidth', 1.5, 'MarkerFaceColor', 'b');
hold on;
yline(90, 'g--', 'Cible 90 dB', 'LineWidth', 1.5);
xlabel('Débit (kbits/s)'); ylabel('SNR (dB)');
title('SNR vs débit (stéréo)');
grid on;

% 3. Signal quantifié pour différents bits
b_examples = [4, 8, 12, 16];
t_plot = t(1:min(200, N));  % Premiers échantillons

for i = 1:length(b_examples)
    subplot(2, length(b_examples), length(b_examples) + i);
    b_ex = b_examples(i);
    x_q_ex = round(x(1:length(t_plot)) * 2^(b_ex-1)) / 2^(b_ex-1);
    e_ex = x(1:length(t_plot)) - x_q_ex;
    
    yyaxis left;
    plot(t_plot*1000, x(1:length(t_plot)), 'b-', 'LineWidth', 0.5);
    hold on;
    plot(t_plot*1000, x_q_ex, 'r-', 'LineWidth', 1);
    ylabel('Amplitude (V)');
    
    yyaxis right;
    plot(t_plot*1000, e_ex, 'g-', 'LineWidth', 0.5);
    ylabel('Erreur (V)');
    
    xlabel('Temps (ms)');
    title(sprintf('b = %d bits, SNR = %.1f dB', b_ex, ...
        10*log10(mean(x(1:length(t_plot)).^2) / max(mean(e_ex.^2), eps))));
    grid on;
    legend('Original', 'Quantifié', 'Erreur', 'Location', 'best', 'FontSize', 6);
end

sgtitle('Simulation de la quantification - Problème III', 'FontSize', 14, 'FontWeight', 'bold');
saveas(gcf, '../figures/simulation_quantification.png');

%% Affichage des résultats numériques
fprintf('--- Résultats ---\n');
fprintf('  b (bits) | SNR mesuré (dB) | SNR théorique (dB) | Débit stéréo (kbits/s)\n');
fprintf('  ---------|-----------------|--------------------|-----------------------\n');
for idx = 1:length(bits_range)
    b = bits_range(idx);
    fprintf('  %8d | %15.2f | %18.2f | %21.2f\n', ...
        b, SNR_mesure(idx), SNR_theorie(idx), debit(idx)/1000);
end

% Trouver le nombre de bits minimum pour 90 dB
b_min = find(SNR_mesure >= 90, 1);
if ~isempty(b_min)
    fprintf('\n  → Nombre de bits minimum pour SNR >= 90 dB : %d bits\n', bits_range(b_min));
    fprintf('    Débit stéréo correspondant : %.2f kbits/s\n', debit(b_min)/1000);
end

fprintf('\nFigure sauvegardée : ../figures/simulation_quantification.png\n');

File: python/simulation_quantification.py
"""
==========================================================================
 SIMULATION_QUANTIFICATION.py - Simulation de la quantification
 APP Signal - ISEP 2025-2026 - Problème III
==========================================================================
"""

import numpy as np
import matplotlib.pyplot as plt
import os


def quantifier(x, b, A=1.0):
    """
    Quantifie un signal sur b bits.
    
    x : signal d'entrée (amplitude entre -A/2 et +A/2)
    b : nombre de bits
    A : dynamique totale
    
    Retourne : signal quantifié, erreur de quantification
    """
    x_q = np.round(x * 2**(b - 1)) / 2**(b - 1)
    e = x - x_q
    return x_q, e


def calcul_snr(x, x_q):
    """Calcule le SNR en dB."""
    e = x - x_q
    P_signal = np.mean(x**2)
    P_erreur = np.mean(e**2)
    if P_erreur > 0:
        return 10 * np.log10(P_signal / P_erreur)
    return np.inf


def simulation_quantification(Fe=44100, V_max=0.5, b_max=20, f0=1000):
    """
    Simule la quantification pour différents nombres de bits.
    
    Fe    : fréquence d'échantillonnage (Hz)
    V_max : amplitude crête (V)
    b_max : nombre max de bits à tester
    f0    : fréquence du signal test (Hz)
    """
    A = 2 * V_max
    duree = 0.1  # secondes
    t = np.arange(0, duree, 1/Fe)
    N = len(t)
    
    # Signal test : sinusoïde
    x = V_max * np.sin(2 * np.pi * f0 * t)
    
    print(f"  Signal test : sinusoïde {f0:.0f} Hz, amplitude {V_max} V")
    print(f"  Fe = {Fe} Hz, durée = {duree} s, N = {N} échantillons")
    
    bits_range = np.arange(1, b_max + 1)
    SNR_mesure = np.zeros(len(bits_range))
    SNR_theorie = np.zeros(len(bits_range))
    debits = np.zeros(len(bits_range))
    
    for idx, b in enumerate(bits_range):
        x_q, e = quantifier(x, b, A)
        
        P_signal = np.mean(x**2)
        P_erreur = np.mean(e**2)
        
        if P_erreur > 0:
            SNR_mesure[idx] = 10 * np.log10(P_signal / P_erreur)
        else:
            SNR_mesure[idx] = 200  # valeur maximale pour l'affichage
        
        q = A / (2**b)
        Pe_th = q**2 / 12
        SNR_theorie[idx] = 10 * np.log10(P_signal / Pe_th)
        
        debits[idx] = Fe * b * 2  # stéréo
    
    # Nombre de bits minimum pour 90 dB
    idx_90 = np.where(SNR_mesure >= 90)[0]
    if len(idx_90) > 0:
        b_min_90 = bits_range[idx_90[0]]
        print(f"\n  Nombre de bits minimum pour SNR ≥ 90 dB : {b_min_90} bits")
        print(f"  Débit stéréo correspondant : {debits[idx_90[0]]/1000:.2f} kbits/s")
    
    # --- AFFICHAGE ---
    fig, axes = plt.subplots(2, 3, figsize=(16, 10))
    fig.suptitle('Simulation de la quantification - Problème III', fontsize=14, fontweight='bold')
    
    # SNR vs bits
    ax1 = axes[0, 0]
    ax1.plot(bits_range, SNR_mesure, 'bo-', linewidth=1.5, markersize=4, label='Mesuré')
    ax1.plot(bits_range, SNR_theorie, 'r--', linewidth=1.5, label='Théorique')
    ax1.axhline(90, color='g', linestyle='--', linewidth=1.5, label='Cible 90 dB')
    ax1.set_xlabel('Nombre de bits (b)')
    ax1.set_ylabel('SNR (dB)')
    ax1.set_title('SNR vs nombre de bits')
    ax1.legend()
    ax1.grid(True)
    
    # SNR vs débit
    ax2 = axes[0, 1]
    ax2.plot(debits / 1000, SNR_mesure, 'bo-', linewidth=1.5, markersize=4)
    ax2.axhline(90, color='g', linestyle='--', linewidth=1.5, label='Cible 90 dB')
    ax2.set_xlabel('Débit (kbits/s)')
    ax2.set_ylabel('SNR (dB)')
    ax2.set_title('SNR vs débit (stéréo)')
    ax2.legend()
    ax2.grid(True)
    
    # Erreur théorique vs mesurée
    ax3 = axes[0, 2]
    ax3.plot(bits_range, SNR_mesure - SNR_theorie, 'ko-', linewidth=1, markersize=3)
    ax3.set_xlabel('Nombre de bits (b)')
    ax3.set_ylabel('Écart SNR (dB)')
    ax3.set_title('Écart mesure - théorie')
    ax3.grid(True)
    
    # Signaux quantifiés pour différents bits
    b_examples = [4, 8, 12, 16]
    n_plot = min(200, N)
    t_plot = t[:n_plot] * 1000  # en ms
    
    for i, b_ex in enumerate(b_examples):
        ax = axes[1, i] if i < 3 else None
        if ax is None:
            break
        
        x_q_ex, e_ex = quantifier(x[:n_plot], b_ex, A)
        snr_ex = calcul_snr(x[:n_plot], x_q_ex)
        
        ax.plot(t_plot, x[:n_plot], 'b-', linewidth=0.5, alpha=0.7, label='Original')
        ax.plot(t_plot, x_q_ex, 'r-', linewidth=1, label='Quantifié')
        ax.set_xlabel('Temps (ms)')
        ax.set_ylabel('Amplitude (V)')
        ax.set_title(f'b = {b_ex} bits, SNR = {snr_ex:.1f} dB')
        ax.legend(fontsize=7)
        ax.grid(True)
    
    plt.tight_layout()
    
    # Sauvegarde
    os.makedirs('../figures', exist_ok=True)
    plt.savefig('../figures/simulation_quantification.png', dpi=150, bbox_inches='tight')
    print(f"\n  Figure sauvegardée : ../figures/simulation_quantification.png")
    
    # Tableau des résultats
    print(f"\n  {'b':>4} | {'SNR mesuré':>12} | {'SNR théorique':>14} | {'Débit stéréo':>16}")
    print(f"  {'':->4}-+-{'':->12}-+-{'':->14}-+-{'':->16}")
    for idx, b in enumerate(bits_range):
        print(f"  {b:4d} | {SNR_mesure[idx]:10.2f} dB | {SNR_theorie[idx]:12.2f} dB | {debits[idx]/1000:12.2f} kb/s")
    
    return bits_range, SNR_mesure, SNR_theorie


if __name__ == "__main__":
    simulation_quantification()
    plt.show()
